allennlp/modules/token_embedders/pretrained_transformer_embedder.py

```diff
@@ -1,41 +1,191 @@
+        that are in reality multiple segments concatenated together, to 2D tensors, i.e.
+
+        [ [CLS] A B C [SEP] [CLS] D E [SEP] ]
+        -> [ [ [CLS] A B C [SEP] ], [ [CLS] D E [SEP] [PAD] ] ]
+        The [PAD] positions can be found in the returned `segment_concat_mask`.
+
+        # Parameters
+
+        token_ids: `torch.LongTensor`
+            Shape: [batch_size, num_segment_concat_wordpieces].
+            num_segment_concat_wordpieces is num_wordpieces plus special tokens inserted in the
+            middle, i.e. the length of: "[CLS] A B C [SEP] [CLS] D E F [SEP]" (see indexer logic).
+        segment_concat_mask: `torch.LongTensor`
+            Shape: [batch_size, num_segment_concat_wordpieces].
+
+        # Returns:
+
+        token_ids: `torch.LongTensor`
+            Shape: [batch_size * num_segments, self._max_length].
+        segment_concat_mask: `torch.LongTensor`
+            Shape: [batch_size * num_segments, self._max_length].
+        """
+        num_segment_concat_wordpieces = token_ids.size(1)
+        num_segments = math.ceil(num_segment_concat_wordpieces / self._max_length)
+        padded_length = num_segments * self._max_length
+        length_to_pad = padded_length - num_segment_concat_wordpieces
+
+        def fold(tensor):  # Shape: [batch_size, num_segment_concat_wordpieces]
+            # Shape: [batch_size, num_segments * self._max_length]
+            tensor = F.pad(tensor, [0, length_to_pad], value=0)
+            # Shape: [batch_size * num_segments, self._max_length]
+            return tensor.reshape(-1, self._max_length)
+
+        return fold(token_ids), fold(segment_concat_mask)
+
+    def _unfold_long_sequences(
+        self, embeddings: torch.FloatTensor, batch_size: int, num_segment_concat_wordpieces: int
+    ) -> torch.FloatTensor:
+        """
+        We take 2D segments of a long sequence and flatten them out to get the whole sequence
+        representation while remove unnecessary special tokens.
+
+        [ [ [CLS]_emb A_emb B_emb C_emb [SEP]_emb ], [ [CLS]_emb D_emb E_emb [SEP]_emb [PAD]_emb ] ]
+        -> [ [CLS]_emb A_emb B_emb C_emb D_emb E_emb [SEP]_emb ]
+
+        We truncate the start and end tokens for all segments, recombine the segments,
+        and manually add back the start tokens. We generally don't need to add back
+        the end tokens -- because the last segment is usually shorter than self._max_length,
+        the final end tokens usually won't be touched in our truncation process.
+
+        There are special cases where sequences have a full last segment. The only issue
+        in this case is that the [SEP] embedding will be wrong. Nevertheless, [SEP] embeddings
+        are rarely, if ever, used, so it shouldn't be a huge problem. However, we do remedy
+        this case when such sequences with a full last segment are the longest ones in a batch
+        by adding back the last token representations.
```

Comment:
I think the reason why there are some issues here is because you're not using the mask associated with the embeddings. If you have the mask it's actually possible to avoid these bugs. For example, if you have the mask, then you know exactly where the [SEP] is for each segment. I'd say it's important to cover these cases robustly. Is there a reason why you avoided using the mask?

Reply:
Discussed offline; will try to do this with fancy index manipulations.